File: problems/0/1/2/012.rs
// Copyright (C) 2014 Jorge Aparicio

use std::iter::count;
use std::mem::replace;
use std::vec_ng::Vec;

fn main() {
    let mut primes = vec!(2);
    let mut next = Factors(vec!((2, 1)));

    for n in count(3u, 1) {
        let curr = replace(&mut next, factorize(n, &mut primes));
        let triangle = next * curr * Factors(vec!((2, -1)));

        if triangle.number_of_divisors() > 500 {
            println!("{}", n * (n - 1) / 2);
            return;
        }
    }
}

// Factors(vec!((a, x), (b, y))) <-> a^x * b^y
struct Factors(Vec<(uint, int)>);

fn factorize(mut n: uint, primes: &mut Vec<uint>) -> Factors {
    let mut factors = vec!();

    for &prime in primes.iter() {
        if n == 1 {
            break;
        }

        if prime * prime > n {
            factors.push((n, 1));
            break;
        }

        let mut i = 0;
        while n % prime == 0 {
            i += 1;
            n /= prime;
        }

        if i != 0 {
            factors.push((prime, i));
        }
    }

    if n != 1 && n > *primes.last().unwrap() {
        primes.push(n);
    }

    Factors(factors)
}

impl Factors {
    fn number_of_divisors(&self) -> int {
        let &Factors(ref s) = self;

        s.iter().fold(1, |n, &(_, e)| n * (e + 1))
    }
}

impl Mul<Factors, Factors> for Factors {
    fn mul(&self, rhs: &Factors) -> Factors {
        let (&Factors(ref s), &Factors(ref r)) = (self, rhs);

        let (m, n) = (s.len(), r.len());
        let (mut i, mut j) = (0 ,0);

        let mut o = vec!();
        while i != m || j != n {
            if i == m {
                o.push(*r.get(j));
                j += 1;
            } else if j == n {
                o.push(*s.get(i));
                i += 1;
            } else {
                let (&(a, x), &(b, y)) = (s.get(i), r.get(j));

                if a > b {
                    o.push((b, y));
                    j += 1;
                } else if a < b {
                    o.push((a, x));
                    i += 1;
                } else {
                    if x + y > 0 {
                        o.push((a, x + y));
                    }

                    i += 1;
                    j += 1;
                }
            }
        }

        Factors(o)
    }
}

File: problems/0/1/2/012.py
from functools import reduce

primes = [2]


def factorize(n):
    factors = []

    for prime in primes:
        if n == 1:
            break

        if prime * prime > n:
            factors.append((n, 1))
            break

        i = 0
        while n % prime == 0:
            i += 1
            n /= prime

        if i != 0:
            factors.append((prime, i))

    if n != 1 and (len(primes) == 0 or n > primes[-1]):
        primes.append(n)

    return factors


def number_of_divisors(factors):
    return reduce(lambda acc, f: acc * (f[1] + 1), factors, 1)


def triangle(left_factors, right_factors):
    (m, n) = (len(left_factors), len(right_factors))
    (i, j) = (0, 0)

    out = []

    while i != m or j != n:
        if i == m:
            out.append(right_factors[j])
            j += 1
        elif j == n:
            out.append(left_factors[i])
            i += 1
        else:
            ((a, x), (b, y)) = (left_factors[i], right_factors[j])

            if a == 2:
                x -= 1
            elif b == 2:
                y -= 1

            if a > b:
                out.append((b, y))
                j += 1
            elif b > a:
                out.append((a, x))
                i += 1
            else:
                if x + y > 0:
                    out.append((a, x + y))

                i += 1
                j += 1

    return out


nod = 0
n = 2
nex = [(2, 1)]
while nod < 500:
    n += 1
    (cur, nex) = (nex, factorize(n))

    nod = number_of_divisors(triangle(cur, nex))

print(int(n * (n - 1) / 2))
